src/mode/modeHandler.ts

```diff
@@ -129,50 +129,63 @@ export class VimSettings {
   useCtrlKeys     = false;
 }
 
+export enum SearchDirection {
+  Forward = 1,
+  Backward = -1
+}
+
 export class SearchState {
+  private static readonly MAX_SEARCH_RANGES = 1000;
+
   /**
    * Every range in the document that matches the search string.
    */
-  public matchRanges: vscode.Range[] = [];
+  private _matchRanges: vscode.Range[] = [];
+  public get matchRanges(): vscode.Range[] {
+    return this._matchRanges;
+  }
+
+  private _searchCursorStartPosition: Position;
+  public get searchCursorStartPosition(): Position {
+    return this._searchCursorStartPosition;
+  }
+
+  private _matchesDocVersion: number;
+  private _searchDirection: SearchDirection = SearchDirection.Forward;
 
   private _searchString = "";
   public get searchString(): string {
     return this._searchString;
   }
 
-  public searchCursorStartPosition: Position;
-
-  /**
-   * 1  === forward
-   * -1 === backward
-   */
-  public searchDirection = 1;
-
-
   public set searchString(search: string){
-    this._searchString = search;
-
-    this._recalculateSearchRanges();
+    if (this._searchString !== search) {
+      this._searchString = search;
+      this._recalculateSearchRanges(/*forceRecalc=*/true);
+    }
   }
 
-  private _recalculateSearchRanges(): void {
+  private _recalculateSearchRanges(forceRecalc?: boolean): void {
     const search = this.searchString;
 
     if (search === "") { return; }
 
-    // Calculate and store all matching ranges
-    this.matchRanges = [];
+    if (this._matchesDocVersion !== TextEditor.getDocumentVersion() || forceRecalc) {
+      // Calculate and store all matching ranges
+      this._matchesDocVersion = TextEditor.getDocumentVersion();
+      this._matchRanges = [];
 
-    for (let lineIdx = 0; lineIdx < TextEditor.getLineCount(); lineIdx++) {
-      const line = TextEditor.getLineAt(new Position(lineIdx, 0)).text;
+      for (let lineIdx = 0; lineIdx < TextEditor.getLineCount() && this._matchRanges.length < SearchState.MAX_SEARCH_RANGES; lineIdx++) {
+        const line = TextEditor.getLineAt(new Position(lineIdx, 0)).text;
 
-      let i = line.indexOf(search);
+        let i = line.indexOf(search);
 
-      for (; i !== -1; i = line.indexOf(search, i + search.length)) {
-        this.matchRanges.push(new vscode.Range(
-          new Position(lineIdx, i),
-          new Position(lineIdx, i + search.length)
-        ));
+        for (; i !== -1 && this._matchRanges.length < SearchState.MAX_SEARCH_RANGES; i = line.indexOf(search, i + search.length)) {
```

Comment:
Instead of duplicating the MAX_SEARCH_RANGES check I would suggest to label the outer for loop with `outer:` and then do something like:

`if (this._matchRanges.length > SearchState.MAX_SEARCH_RANGES) { break outer; }`

within the inner loop.